src/docfx/lib/json/JsonContractResolver.cs

```diff
@@ -29,61 +29,69 @@ protected override JsonObjectContract CreateObjectContract(Type objectType)
             {
-                if (prop.PropertyType.IsGenericType && prop.PropertyType.GetGenericTypeDefinition() == typeof(SourceInfo<>))
+                if (!property.Writable)
                 {
-                    var originalShouldSerialize = prop.ShouldSerialize;
-                    prop.ShouldSerialize =
-                        target =>
-                        {
-                            var value = prop.ValueProvider.GetValue(target);
-                            if (((SourceInfo)value)?.GetValue() is null)
-                                return false;
-
-                            if (IsEmptyArray(((SourceInfo)value).GetValue()))
-                            {
-                                return false;
-                            }
-                            return originalShouldSerialize?.Invoke(target) ?? true;
-                        };
+                    if (member is FieldInfo f && f.IsPublic && !f.IsStatic)
+                    {
+                        property.Writable = true;
+                    }
                 }
             }
+        }
 
-            void ShouldNotSerializeEmptyArray()
+        private void DoNotSerializeEmptyArray(JsonProperty property)
+        {
+            if (typeof(IEnumerable).IsAssignableFrom(property.PropertyType) && !(property.PropertyType == typeof(string)))
             {
-                if (typeof(IEnumerable).IsAssignableFrom(prop.PropertyType) && !(prop.PropertyType == typeof(string)))
-                {
-                    prop.ShouldSerialize =
-                        target => !IsEmptyArray(prop.ValueProvider.GetValue(target));
-                }
-            }
+                var originalShouldSerialize = property.ShouldSerialize;
 
-            bool IsEmptyArray(object value)
-            {
-                if (value is IEnumerable enumer && !enumer.GetEnumerator().MoveNext())
+                property.ShouldSerialize = target =>
                 {
-                    return true;
-                }
-                return false;
+                    if (IsEmptyArray(property.ValueProvider.GetValue(target)))
+                    {
+                        return false;
+                    }
+
+                    return originalShouldSerialize?.Invoke(target) ?? true;
+                };
             }
+        }
 
-            void SetFieldWritable()
+        private static bool IsEmptyArray(object value)
+        {
+            return !(value is string) && value is IEnumerable enumerable && !enumerable.GetEnumerator().MoveNext();
+        }
+
+        private static void HandleSourceInfo(JsonProperty property)
+        {
+            if (property.PropertyType.IsGenericType && property.PropertyType.GetGenericTypeDefinition() == typeof(SourceInfo<>))
             {
-                if (!prop.Writable)
+                // Allow source info propagation for null values
+                property.NullValueHandling = NullValueHandling.Include;
```

Comment:
keep the behavior consistent with normal type null handling: null value means using docfx default value.